README.md
# @mconnect/mcdb

- Db Connection package for mongoDB | PostgreSQL/SQLite3 (sequelize)


HISTORY.md
# Historical logs

## Version 0.2.7
- Release date: Jun 06, 2021
- updated mcresponse package to v0.4.1

## Version 0.2.6
- Release date: Jun 06, 2021
- updated mcresponse package to v0.4.1

## Version 0.2.5

- Release date: Jan 25, 2021
- resolved mongo-dBase authentication error

## Version 0.2.4

- Release date: Jan 24, 2021
- resolved mongo-dBase options type and auth errors

## Version 0.2.3

- Release date: Dec 14, 2020
- resolved types link in package.json

## Version 0.2.2

- Release date: Dec 14, 2020
- resolved types definition packaging issue

## Version 0.2.1

- Release date: Dec 14, 2020
- added the validator package for sequelize

## Version 0.2.0

- Release date: Dec 14, 2020
- npm/github publication: refactor, test, build and release

## Version 0.1.0

- Release date: July 14, 2020
- Db Connect module: sequelize...


test/dbTask.ts
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2020-07-17
 * @Company: Copyright 2020 Abi Akindele  | mConnect.biz
 * @License: All Rights Reserved | LICENSE.md
 * @Description: mc-central-ts: mongodb testing
 */

const collRecords: any = {
    name    : "Abi",
    desc    : "Testing only",
    url     : "localhost:9000",
    priority: 1,
    cost    : 1000.00
};
const newCollRecords: any = {
    name    : "Abi Akindele",
    desc    : "Testing only - updated",
    url     : "localhost:9900",
    priority: 1,
    cost    : 2000.00
};

export const dbTaskRecord = {
    collName   : 'users',
    collRecords: collRecords,
    logType    : 'create',
    logBy      : 'abbeycity-testing-db-task',
    logDate    : new Date(),
};


test/db.test.ts
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2020-12-14
 * @Company: Copyright 2020 Abi Akindele  | mConnect.biz
 * @License: All Rights Reserved | LICENSE.md
 * @Description: SQL-DB Connection test cases
 */

import { mcTest, assertEquals, postTestResult } from "@mconnect/mctest";
import {checkDb, DbConfigType} from "../src";
import { newDbConnect } from "../src";

// test config/data

// test-data: db-configuration settings
const myDb = {
    host    : "localhost",
    username: "postgres",
    password: "ab12testing",
    port    : 5432,
    dbName  : "mcdev",
    filename: "testdb.db",
    poolSize: 20,
    secureOptions: {
        secureAccess: false,
        secureCert: "",
        secureKey: "",
    },
    uri     : "localhost:5432",
    options : {},
};

const sqliteDb = {
    filename: "testdb.db",
};

(async () => {
    // pre-testing setup
    const dbc = newDbConnect(myDb, "postgres");
    const dbOpen = await dbc?.openDb();

    const dbc2 = newDbConnect(sqliteDb, "sqlite");
    const dbOpen2 = await dbc?.openDb();

    // perform db-connection testing
    await mcTest({
        name    : "should successfully connect to the PostgreSQL DB ",
        testFunc: async () => {
            const dbRes = checkDb(dbOpen)
            assertEquals(dbRes.code, "success", "response-code should be: success")
            assertEquals(dbRes.message.includes("valid database connection/handler"), true, "response-message should be: true")
        },
    });

    await mcTest({
        name    : "should successfully connect to the SQLite3 DB ",
        testFunc: async () => {
            const dbRes = checkDb(dbOpen2)
            assertEquals(dbRes.code, "success", "response-code should be: success")
            assertEquals(dbRes.message.includes("valid database connection/handler"), true, "response-message should be: true")
        },
    });

    postTestResult()

    // close resources / avoid memory leak
    await dbc?.closeDb();
    await dbc2?.closeDb();
})()




test/dbError.test.ts
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2020-12-14
 * @Company: Copyright 2020 Abi Akindele  | mConnect.biz
 * @License: All Rights Reserved | LICENSE.md
 * @Description: SQL-DB Connection test cases - error testing | TODO: review error-test-cases
 */

import { mcTest, assertEquals, postTestResult } from "@mconnect/mctest";
import {checkDb, DbConfigType} from "../src";
import { newDbConnect } from "../src";
import {dbs} from "./config/secure/dbConfig";

// test-data: db-configuration settings
const myDb = {
    host    : "localhost",
    username: "postgres",
    password: "ab12testing2",
    port    : 5400,
    dbName  : "mcdev-null",
    filename: "",
    poolSize: 20,
    secureOptions: {
        secureAccess: false,
        secureCert: "",
        secureKey: "",
    },
    uri     : "localhost:5400",
    options : {},
};

const sqliteDb = {
    filename: "",
};

(async () => {
    // pre-testing setup
    const dbc = newDbConnect(dbs.postgres, "postgres");
    const dbOpen = await dbc?.openDb();

    const dbc2 = newDbConnect(dbs.sqlite, "sqlite");
    const dbOpen2 = await dbc?.openDb();

    // perform db-connection testing
    await mcTest({
        name    : "should report connection-error to the PostgreSQL DB (dv-name-error)",
        testFunc: async () => {
            console.log("pg-db: ", dbOpen)
            const dbRes = checkDb(dbOpen)
            assertEquals(dbRes.code, "validateError", "response-code should be: validateError")
            assertEquals(dbRes.message.includes("valid database connection/handler is required"), true, "response-message should be: true")
        },
    });

    await mcTest({
        name    : "should report connection-error to the SQLite3 DB (filename-error)",
        testFunc: async () => {
            console.log("sqlite-db: ", dbOpen2)
            const dbRes = checkDb(dbOpen2)
            assertEquals(dbRes.code, "validateError", "response-code should be: validateError")
            assertEquals(dbRes.message.includes("valid database connection/handler is required"), true, "response-message should be: true")
        },
    });

    postTestResult()

    // close resources / avoid memory leak
    await dbc?.closeDb();
    await dbc2?.closeDb();
})()




src/DbMongo.ts
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2020-07-17
 * @Company: Copyright 2020 Abi Akindele  | mConnect.biz
 * @License: All Rights Reserved | LICENSE.md
 * @Description: mc: db connection for mongoDB
 */

import { Db, MongoClient } from "mongodb";
import { DbConfigType, DbSecureType, MongoDbOptionsType } from "./types";

export class DbMongo {
    private readonly host: string;
    private readonly username: string;
    private readonly password: string;
    private readonly database: string;
    private readonly location: string;
    private readonly port: number;
    private readonly poolSize: number;
    private readonly secureOption: DbSecureType;
    private serverUrl: string;
    private dbUrl: string;
    private readonly options: MongoDbOptionsType;
    private readonly checkAccess: boolean;
    private readonly user: string;
    private readonly pass: string;
    private mcDb: any;

    constructor(dbConfig: DbConfigType, options?: MongoDbOptionsType) {
        this.host = dbConfig?.host || "";
        this.username = dbConfig?.username || "";
        this.password = dbConfig?.password || "";
        this.database = dbConfig?.database || "";
        this.location = dbConfig?.location || "";
        this.port = Number(dbConfig?.port) || Number.NEGATIVE_INFINITY;
        this.poolSize = dbConfig?.poolSize || 20;
        this.secureOption = dbConfig?.secureOption || {secureAccess: false, secureCert: "", secureKey: ""};
        this.checkAccess = true;
        if (options?.checkAccess === false) {
            this.checkAccess = false;
        }
        this.user = encodeURIComponent(this.username);
        this.pass = encodeURIComponent(this.password);
        // this.user = this.username;
        // this.pass = this.password;
        this.dbUrl = this.checkAccess ? `mongodb://${this.user}:${this.pass}@${dbConfig.host}:${dbConfig.port}/${dbConfig.database}` :
            `mongodb://${dbConfig.host}:${dbConfig.port}/${dbConfig.database}`;
        this.serverUrl = this.checkAccess ? `mongodb://${this.user}:${this.pass}@${dbConfig.host}:${dbConfig.port}` :
            `mongodb://${dbConfig.host}:${dbConfig.port}`;
        this.options = {
            poolSize          : options?.poolSize || this.poolSize,
            // reconnectTries    : options?.reconnectTries || Number.MAX_VALUE,
            // reconnectInterval : options?.reconnectInterval || 1000,
            useNewUrlParser   : options?.useNewUrlParser || true,
            useUnifiedTopology: options?.useUnifiedTopology || true,
        };
    }

    async connectServer() {
        try {
            return await this.mgServer();
        } catch (e) {
            throw new Error("MongoDB server connection error:" + e.message);
        }
    }

    async openDb(dbName = "") {
        try {
            return await this.mgDb(dbName);
        } catch (e) {
            throw new Error("MongoDB opening error:" + e.message);
        }
    }

    async closeDb() {
        if (this.mcDb) {
            await this.mcDb?.close();
        }
    }

    async mgServer(): Promise<MongoClient> {
        const dbenv = process.env.NODE_ENV || "development";
        if (dbenv === "production" && process.env.MONGODB_URI) {
            this.dbUrl = process.env.MONGODB_URI;
        }
        try {
            const client = new MongoClient(this.dbUrl, this.options);
            this.mcDb = await client.connect();
            return this.mcDb;
        } catch (err) {
            if (this.mcDb) {
                await this.mcDb.close();
            }
            console.error("MongoDB server connection error:" + err.stack);
            throw new Error("MongoDB server connection error:" + err.message);
        }

    }

    async mgDb(dbName = ""): Promise<Db> {
        let client: MongoClient;
        try {
            // connect to the server (pool connections)
            client = await this.mgServer();
            return client.db(dbName || this.database);
        } catch (err) {
            if (this.mcDb) {
                await this.mcDb.close();
            }
            console.error("MongoDB connection error:" + err.stack);
            throw new Error("Error opening/creating a mongo database handle | " + err.message);
        }
    }
}

export function newDbMongo(dbConfig: DbConfigType, options?: MongoDbOptionsType) {
    return new DbMongo(dbConfig, options);
}


src/DbConnect.ts
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2020-07-14
 * @Company: Copyright 2020 Abi Akindele  | mConnect.biz
 * @License: All Rights Reserved | LICENSE.md
 * @Description: mc-central-ts: Db-connect class for sequelize package
 */

import { DbConfigType, DbSecureType, DbType, DbConnectOptions, DbConnectType } from "./types";
import { Sequelize } from "sequelize";

export class DbConnect {
    private readonly host: string;
    private readonly username: string;
    private readonly password: string;
    private readonly database: string;
    private readonly filename: string;
    private readonly location: string;
    private readonly port: number;
    private readonly dbType: DbType;
    private readonly poolSize: number;
    private readonly secureOption: DbSecureType;
    private readonly uri: string;
    private mcDb: DbConnectType;
    private readonly connectionPoolOptions: object;

    constructor(dbConfig: DbConfigType, dbType: DbType = "postgres", options: DbConnectOptions = {}) {
        this.host = dbConfig?.host || "";
        this.username = dbConfig?.username || "";
        this.password = dbConfig?.password || "";
        this.database = dbConfig?.database || "";
        this.filename = dbConfig?.filename || "";
        this.location = dbConfig?.location || "";
        this.port = Number(dbConfig?.port) || Number.NEGATIVE_INFINITY;
        this.dbType = dbType;
        this.poolSize = dbConfig?.poolSize || 20;
        this.secureOption = dbConfig?.secureOption || {secureAccess: false, secureCert: "", secureKey: ""};
        this.uri = dbConfig?.uri || "";
        this.mcDb = new Sequelize({dialect: "postgres"});
        this.connectionPoolOptions = options || {
            max    : 20,
            min    : 0,
            acquire: 30000,
            idle   : 10000,
        };
    }

    async openDb(): Promise<DbConnectType> {
        await this.dbConnect();
        return this.mcDb;
    }

    async closeDb() {
        await this.mcDb?.close();
    }

    async dbConnect(): Promise<DbConnectType> {
        try {
            switch (this.dbType) {
                case "sqlite":
                    this.mcDb = new Sequelize({
                        dialect: "sqlite",
                        storage: this.location, /*"path/to/database.sqlite"*/
                    });
                    break;
                case "postgres":
                    // dbURI = `postgres://${this.username}:${this.password}@${this.host}:${this.port}/${this.database}`;
                    this.mcDb = new Sequelize(this.database, this.username, this.password, {
                        host   : this.host,
                        port   : this.port,
                        dialect: this.dbType, /* one of "mysql" | "mariadb" | "postgres" | "mssql" */
                        pool   : this.connectionPoolOptions,
                    });
                    break;
                case "mariadb":
                    this.mcDb = new Sequelize(this.database, this.username, this.password, {
                        host   : this.host,
                        port   : this.port,
                        dialect: this.dbType, /* one of "mysql" | "mariadb" | "postgres" | "mssql" */
                        pool   : this.connectionPoolOptions,
                    });
                    break;
                case "mysql":
                    this.mcDb = new Sequelize(this.database, this.username, this.password, {
                        host   : this.host,
                        port   : this.port,
                        dialect: this.dbType, /* one of "mysql" | "mariadb" | "postgres" | "mssql" */
                        pool   : this.connectionPoolOptions,
                    });
                    break;
                case "mssql":
                    this.mcDb = new Sequelize(this.database, this.username, this.password, {
                        host   : this.host,
                        port   : this.port,
                        dialect: this.dbType, /* one of "mysql" | "mariadb" | "postgres" | "mssql" */
                        pool   : this.connectionPoolOptions,
                    });
                    break;
                default:
                    throw new Error("unsupported database type: " + this.dbType);
                    break;
            }
            return this.mcDb;
        } catch (e) {
            console.error("error starting DB server: ", e.message, e);
            throw new Error("error starting DB server: " + e.message);
        }
    }
}

export function newDbConnect(dbConfig: DbConfigType, dbType: DbType = "postgres", options: DbConnectOptions = {}) {
    return new DbConnect(dbConfig, dbType, options);
}


src/types.ts
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2020-07-14
 * @Company: Copyright 2020 Abi Akindele  | mConnect.biz
 * @License: All Rights Reserved | LICENSE.md
 * @Description: mc-central-ts: db types for sequelize package
 */

import { Sequelize } from "sequelize";
import { Db } from "mongodb";

// types
export interface DbSecureType {
    secureAccess: boolean;
    secureCert?: string;
    secureKey?: string;
}

export type DbConnectType = Sequelize;

export type MongoDbConnectType = Db;

export type DbType = "postgres" | "sqlite" | "mysql" | "mariadb" | "mssql";

export interface DbConfigType {
    host?: string;
    username?: string;
    password?: string;
    database?: string;
    filename?: string;
    location?: string;
    port?: number | string;
    poolSize?: number;
    secureOption?: DbSecureType;
    uri?: string;
}

export interface DbConfigsType {
    [key: string]: DbConfigType;
}

export interface DbConnectOptions {
    [key: string]: string | number | object | boolean;
}

export interface MongoDbOptionsType {
    checkAccess?: boolean;
    poolSize?: number;
    reconnectTries?: number;
    reconnectInterval?: number;
    useNewUrlParser?: boolean;
    useUnifiedTopology?: boolean;
}

export interface DatabaseType {
    dbType: DbType;
    postgres: DbConfigsType;
    mongodb: DbConfigsType;
    redis: DbConfigsType;
    mysql: DbConfigsType;
    mariadb: DbConfigsType;
    mssql: DbConfigsType;
    sqlite: DbConfigsType;
}

export interface RoleServiceType {
    serviceId: string;
    groupId: string;
    serviceCategory: string;
    canRead: boolean;
    canCreate: boolean;
    canUpdate: boolean;
    canDelete: boolean;
    collAccessPermitted?: boolean;
}

export interface CheckAccessType {
    userId: string;
    userRole: string;
    userRoles: Array<string>;
    isActive: boolean;
    isAdmin: boolean;
    roleServices: Array<RoleServiceType>;
    collId: string;
}

export interface OkResponseType {
    ok: boolean;
}

export interface RoleFuncType {
    (it1: string, it2: RoleServiceType): boolean;
}

// Exception/error types
export type SaveError = Error;
export type CreateError = Error;
export type UpdateError = Error;
export type DeleteError = Error;
export type ReadError = Error;
export type AuthError = Error;
export type ConnectError = Error
export type SelectQueryError = Error
export type WhereQueryError = Error
export type CreateQueryError = Error
export type UpdateQueryError = Error
export type DeleteQueryError = Error
